### PR TITLE
Core polish

### PR DESCRIPTION
**Pull request by angeredShrimp**

### CODE REVIEW

File: kernel/core/extend/intr.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            debug("phys ->%p\n", (void*)phys);
+
+            // Trick for compiling
+            // TODO
+            uint32_t a;
+            a = (uint32_t)phys;
+            a = (uint32_t)tidx;
+            a = (uint32_t)task_PGD;
+            a = (uint32_t)task_PTB;
+            a = a;
```
Comment (Zefinder) [inline kernel/core/extend/intr.c:47]: What is that...

File: kernel/core/extend/pagemem.c [hunk 2]
```diff
@@ -17,33 +17,37 @@ void init_pgd()
...
 	/*********************  TODO: Should ultimately be kernel mapped  *********************/
 	// Mapping user memory
-	usr_identity_map(&PGD[0], &PTB[0], USER_START, USER_END); // USR
+	usr_identity_map(&PGD[0], USER_START, USER_END); // USR
 
 	// Mapping userland, for now userland is in Kernel memory
-	usr_identity_map(&PGD[0], &PTB[0], (offset_t)userland, (offset_t)userland + PDE_OFFSET - 1);
-	usr_identity_map(&PGD[0], &PTB[0], (offset_t)printf, (offset_t)userland + PDE_OFFSET - 1);
+	usr_identity_map(&PGD[0], (offset_t)userland, (offset_t)userland + PTE_OFFSET - 1);
+	usr_identity_map(&PGD[0], (offset_t)printf, (offset_t)userland + PTE_OFFSET - 1);
```
Comment (Zefinder) [inline kernel/core/extend/pagemem.c:47]: Yes, we need to find a way to put user functions somewhere... copy? Must be next feature!

File: kernel/core/extend/stacks.c [hunk 3]
```diff
@@ -15,18 +21,45 @@ void init_task_stack() {
...
     // Else we return the task index value and we increase the task_stack_index
     return task_stack_index == TASK_NUMBER ? -1 : task_stack[task_stack_index++];
-}
\ No newline at end of file
+}
+
+// Pops the ptb that is ready to use, removing it from the stack. If none returns -1
+pte32_t* pop_ptb() {
+    // If no value is pointed (after NUM_PTB - 1), we return -1
+    // Else we return the task index value and we increase the task_stack_index
```
Comment (Zefinder) [inline kernel/core/extend/stacks.c]: Why do you return the task index 😭

File: kernel/core/extend/userland.c [hunk 4]
```diff
@@ -1,12 +1,21 @@
...
 void userland() {
    debug("This is fine.\n");
 
+   // Syscall
+   /*
+   ** TODO: change asm automatic
+   */
+   // asm volatile ("mov $1, %eax ; int $128\n\t");
+   uint32_t test = 42;
+   asm volatile ("mov %0, %%eax ; mov %1, %%ebx ; int %2\n\t" :: "i" (SYS_READ_CNT),"i" () , "i" (INT_SYSCALL) : "%eax", "%ebx");
```
Comment (Zefinder) [inline kernel/core/extend/userland.c]:
```suggestion
   int read_cnt = SYS_READ_CNT;
   int val = VALUE;
   asm volatile ("int %2\n\t" :: "a" (read_cnt), "b" (val) , "i" (INT_SYSCALL));
```
No need for clobbers here. You use clobbers when you use a register without declaring it in input or output

File: kernel/include/extend/memory.h [hunk 5]
```diff
@@ -2,6 +2,7 @@
 #define MEMORY_H
```
Comment (Zefinder) [inline kernel/include/extend/memory.h]: Don't forget to change every define headers if you use `__EXTENDS_NAME_H__`

File: kernel/include/extend/pagemem.h [hunk 6]
```diff
@@ -51,146 +80,86 @@
...
-		for(offset_t p = _start_; p <= _end_; p += PTE_OFFSET) {				\
-			ptb_set_krn_identity_pte( ((pde32_t*)(_ptbs_)), p );				\
-		}																		\
+#define krn_identity_map(_pgd_,_start_,_end_)													\
+	{																							\
+		pde32_t *pde;																			\
+		pte32_t *ptb, *pte;																		\
+		for(offset_t p = (offset_t)(_start_); p <= (offset_t)(_end_); p += PDE_OFFSET) {		\
+			pde = &(_pgd_)[pd32_idx(p)];														\
+			ptb = pde->p ? (pte32_t*)(pde->addr << 12) : pop_ptb();								\
```
Comment (Zefinder) [inline kernel/include/extend/pagemem.h:114]: readablen't but it seems correct

**Review by Zefinder — COMMENTED** [review]
That's really good progress!

File: kernel/core/extend/intr.c (hunk 1 above)
Comment (angeredShrimp) [reply kernel/core/extend/intr.c:47]: Trick for compiling during progress as compiler checks if variables are used (in the right hand side of affectation)

File: kernel/core/extend/pagemem.c (hunk 2 above)
Comment (angeredShrimp) [reply kernel/core/extend/pagemem.c:47]: By defining a .user section in linker script and adding a section attribute to the functions!

File: kernel/core/extend/stacks.c (hunk 3 above)
Comment (angeredShrimp) [reply kernel/core/extend/stacks.c]: Woooops

**Review by Zefinder — COMMENTED** [review]
Add test for pte stack

File: unit_test/test_script.sh
Comment (Zefinder) [inline unit_test/test_script.sh:1]: Interesting

**Review by Zefinder — APPROVED** [review]
This can be merged! Don't forget to pull on main and rebase on other working branches after merging

**Pull request merged**